enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+API endpoint can be used.
+
+Using the [horizontal pod
+autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+as a reference, they are using the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
+to gather performance statistics for nodes. 
+
+Using the metrics API has the least requirements as it does not expect a running
+Prometheus monitoring stack. Using Prometheus generated data could make the
+implementation easier with respect to data aggregation.
+
+Data aggregation will be required using the metrics API, as decisions about
+scaling nodes up and down, must not be performed based on performance usage at a
+single point in time. Instead the operator will have to gather and store usage
+data for a certain period of time, to decide if scaling is required.
+
+#### Hypershift
+
+Hypershift does not use `ControlPlaneMachineSet` to manage the control-planes
+for hosted clusters, as those are not running on their own virtual machines.
+
+Because of this, no special handling should be required.
+
+### Risks and Mitigations
+
+#### Risk 1: Higher costs when scaling up
+
+Scaling up will increase costs for running the cluster, as bigger nodes will
+incur higher costs from the cloud provider.
+
+#### Risk 2: Cluster instability due to scaling down
+
+Scaling down might also impact the clusters ability to accommodate a workload if
+resource usage was uncharacteristically low for the detection period of scaling.
+This turns into an issue, if the cluster still would have required the bigger
+control plane nodes.
+
+To mitigate this issue the scale down behavior should be disabled by default and
+require an explicit opt-in by the cluster administrator.
+
+#### Risk 3: Control plane node churn
+
+Automatically scaling up and down control planes could impact cluster
+reliability and availability, if performed too often.
+
+As such the additional logic must ensure that there is a grace period between
+changes to the machine size, during which no more scaling will be performed.
+
+### Drawbacks
+
+Some users might not want this feature and rather opt to scale control planes
+manually.
+
+## Design Details
+
+### Control Plane MachinSet operator configuration
+
+As automatic scaling will require adjustments to certain aspects of its logic,
+the implementation should implement a way for administrators of an OpenShift
+cluster to configure the specifics of how and when scaling will occur.
+
+There should be certain **must** configurations that have to be filled out
+explicitly, while other are **can** configurations that should come with
+sensible defaults.
+
+The **must** configurations will include the following properties:
+
+- `instancesSizes`: a yaml object defining, the available machine sizes and an
+  ordering to determine the order in which to size up and down.
+- `instanceSizePath`: the path of the `instanceType` in the
```

Comment:
Not every cloud uses a single field for this, how will you account for vSphere for example?